sw/airborne/modules/ctrl/optical_flow_landing.c

```diff
@@ -171,434 +195,326 @@ void vertical_ctrl_module_run(bool in_flight);
+    if (of_landing_ctrl.agl_lp > 0.0001f) {
+      of_landing_ctrl.divergence = of_landing_ctrl.vel / of_landing_ctrl.agl_lp;
+      // TODO: this time scaling should be done in optical flow module
+      divergence_vision_dt = (divergence_vision / dt);
+      if (fabsf(divergence_vision_dt) > 1E-5) {
+        div_factor = of_landing_ctrl.divergence / divergence_vision_dt;
       }
-      // reset dt:
-      dt = 0.0f;
     }
   } else {
-
     // USE REAL VISION OUTPUTS:
+    // TODO: this div_factor depends on the subpixel-factor (automatically adapt?)
+    // TODO: this factor is camera specific and should be implemented in the optical
+    // flow calculator module not here. Additionally, the time scaling should also
+    // be done in the calculator module
+    div_factor = -1.28f; // magic number comprising field of view etc.
+    float new_divergence = (divergence_vision * div_factor) / dt;
+
+    // deal with (unlikely) fast changes in divergence:
+    static const float max_div_dt = 0.20;
+    if (fabsf(new_divergence - of_landing_ctrl.divergence) > max_div_dt) {
+      if (new_divergence < of_landing_ctrl.divergence) { new_divergence = of_landing_ctrl.divergence - max_div_dt; }
+      else { new_divergence = of_landing_ctrl.divergence + max_div_dt; }
+    }
 
-    if (vision_message_nr != previous_message_nr && dt > 1E-5 && ind_hist > 1) {
-
-      float new_divergence = (divergence_vision * of_landing_ctrl.div_factor) / dt;
+    // low-pass filter the divergence:
+    of_landing_ctrl.divergence += (new_divergence - of_landing_ctrl.divergence) * lp_factor;
+    prev_vision_time = vision_time;
+  }
 
-      // deal with (unlikely) fast changes in divergence:
-      float max_div_dt = 0.20;
-      if (fabs(new_divergence - divergence) > max_div_dt) {
-        if (new_divergence < divergence) { new_divergence = divergence - max_div_dt; }
-        else { new_divergence = divergence + max_div_dt; }
+  /***********
+  * CONTROL
+  ***********/
+  // landing indicates whether the drone is already performing a final landing procedure (flare):
+  if (!landing) {
+    if (of_landing_ctrl.CONTROL_METHOD == 0) {
+      // FIXED GAIN CONTROL, cov_limit for landing:
+
+      // use the divergence for control:
+      thrust_set = PID_divergence_control(of_landing_ctrl.divergence_setpoint, of_landing_ctrl.pgain, of_landing_ctrl.igain,
+                                          of_landing_ctrl.dgain, dt);
+
+      // trigger the landing if the cov div is too high:
+      if (fabsf(cov_div) > of_landing_ctrl.cov_limit) {
+        thrust_set = final_landing_procedure();
       }
-
-      // low-pass filter the divergence:
-      divergence = divergence * of_landing_ctrl.lp_factor + (new_divergence * (1.0f - of_landing_ctrl.lp_factor));
-      previous_message_nr = vision_message_nr;
-      dt = 0.0f;
-
-    } else {
-      // after re-entering the module, the divergence should be equal to the set point:
-      if (ind_hist <= 1) {
```

Comment:
Was removed. 

It was not clear to me why the divergence should be equal to the set point for one time step when re-entering the module, it would seem that setting the divergence to the set point then back to the real divergence (which would be close to 0 at the beginning of the landing) would not be good for the covariance estimation. 

Advise if that was not the idea.